fix ProductVersion file attribute

enricosada opened the pull request:
fix #605

ProductVersion should be set using the following fallback
sequence:

1. AssemblyInformationalVersionAttribute
2. AssemblyFileVersionAttribute
3. AssemblyVersionAttribute

I changed visibility of some functions from private to internal because i needed to unit test the functions ( fsharp.compiler has already internalvisibleto attribute to fsharp.compiler.tests).

enricosada commented:
This commit also add a way to test the compiler as function, without invoke the command line or spawn another process
I needed a working example to show the test usage ( instead of using the fsharp or fsharpqa suite ). 
If no good, i can remove the test and add another fsharpqa/fsharp test

Note: now the helper function write the output of compiler to filesystem in a temp directory, but that shouldn't be needed if we add a dummy inmemory filesystem

is really pretty simple to add tests:

```fsharp
    [<Test>] 
    member this.``should compile library without warnings`` () = 
        let code =
            """
    module test
    let add x y = x + y
            """
        let r, o = ``fsc_library`` [] ("add.fs", code)
        r |> Check.areEqual 0
        o.outputfile.Exists |> Check.areEqual true
        (o.warnings, o.consoleout) |> Check.areEqual ([], [])
```

the good parts are

```fsharp
    type FscOutput = { 
        warnings: string list;
        consoleout: string list;
        outputfile: FileInfo; }

    let ``fsc`` (args: string list) (name: string, code: string)  -> int (*exitcode*) * FscOutput
    let ``fsc_library`` args input = fsc ("--target:library" :: args) input
```

latkin commented on src/fsharp/FSharp.Compiler.Unittests/FscHelper.fs:
```diff
@@ -0,0 +1,122 @@
+        | v -> failtestf "no invocations expected but were %d: '%A'" v.Length v
+
+
+module ``fsharp compiler`` =
```
Please use standard names so that backticks aren't required

latkin commented on src/fsharp/FSharp.Compiler.Unittests/FscBuildLibrary.fs:
```diff
@@ -0,0 +1,22 @@
+﻿namespace FSharp.Compiler.Unittests.Fsc
```
If this file contains smoke tests, it should have a name representing that -- "FscSmokeTests.fs" or something. This is not a build library so "FscBuildLibrary" doesn't make sense to me.

latkin commented on src/fsharp/FSharp.Compiler.Unittests/FscHelper.fs:
```diff
@@ -0,0 +1,122 @@
+        (sprintf "-o:%s" (outPath |> quote)) :: args @ [sourcefile]
+        |> fscmainwrapper
+
+    let ``fsc_library`` args input = fsc ("--target:library" :: args) input
```
Why does this have backticks?

latkin commented on src/fsharp/FSharp.Compiler.Unittests/FscHelper.fs:
```diff
@@ -0,0 +1,122 @@
+        consoleout: string list;
+        outputfile: FileInfo; }
+
+    let ``fsc`` (args: string list) (name, code) =
```
Why backticks?

enricosada replied:
I used backticks to make easier to understand is not the real `fsc` function. Failed. I'll use standard Module.function

latkin commented:
Have you done testing around the auto-increment version format, with `*`, i.e. stuff like `"1.2.*"`. I forget at what level that is handled, but figured it would be related to this.

latkin commented on src/fsharp/FSharp.Compiler.Unittests/FscHelper.fs:
```diff
@@ -0,0 +1,122 @@
+                let error = new StreamWriter(Console.OpenStandardError()) in error.AutoFlush <- true
+                Console.SetOut(error) }
+
+    let logtomemory () = 
```
Typical F# style (used everywhere else in this repo) would be `logToMemory` camel cased (similarly throughout this PR).  It's not a huge deal, but I don't see any compelling reason to pick a new style just for these tests.

enricosada replied:
Usually i use camel case style. This slipped my review, i'll change it

latkin commented:
Some of the test cases make sense as unit tests, but many are just end-to-end compiler tests. Why not add those with all of the other end-to-end AssemblyVersion tests at `tests\fsharpqa\Source\Conformance\DeclarationElements\CustomAttributes\AttributeUsage`?  

And FWIW the fsharpqa tests already run using and in-memory compiler API. See `tests\fsharpqa\testenv\src\FSharp.Compiler.Hosted`.  IMO it would be better to build out this API than to re-implement it.

enricosada commented:
I think `1.2.*` is resolved before, good case anyway, i'll add a test 

About test:
I know i can add tests in fsharpqa, i'll do it. This is the last time i try to do something different that fsharpqa or fsharp 
So why: because fsharpqa is difficult to use/debug. As i wrote before is lot of times easier use nunit and the nunit runner *inside* vs to run/debug. 
Just try it. Build, test runner, click dx, debug. Now i can debug fsc or the test. 
Like every other tests. 
Using the fcs work in @dsyme prs we can use fake ifilesystem to never hit the disk. 

This is an example, a working use case of end to end test to discuss about the test suite. It can be improved. 

If not good enough i'll remove it and use fsharpqa.

enricosada commented:
@latkin, if you can we can chat about it in  gitter, is faster maybe

dsyme commented on src/fsharp/FSharp.Compiler.Unittests/FscHelper.fs:
```diff
@@ -0,0 +1,122 @@
+
+        let parseWarnings = List.filter (fun (x: string) -> x.StartsWith("warning"))
+
+        let fscmainwrapper a =
```
Another camelCase

latkin commented:
@enricosada there is no good answer, I'm sorry to say. You approach of using unit-style compiler API has many benefits, and I think it's the long-term place we want to get to. I don't want to come across as totally against this idea (though I fear I already do!). I've tried to capture the requirements needed for this in my [comment](https://github.com/Microsoft/visualfsharp/pull/90#issuecomment-127693061) from earlier.

Right now, though, I think it is more prudent to extend the existing tests (in their existing framework) in a sensible way to cover the new cases.

We should try to avoid http://xkcd.com/927/ -style creeping expansion of test frameworks.  The fsharpqa suite was created as an "improvement" over the fsharp suite, and indeed it has some advantages, but it adds a ton of confusion and maintenance problems.

enricosada commented:
I understand the points/requirements in your comment. 
I know is not a final design, as i wrote i want to improve it,
but somewhere we should start. 

Let's set some points ( nunit? Inmemory? ) on a Wiki/issue and iterate on the implementation using use one or two test as use case, not much to lose ( i can duplicate test in fsharpqa so we can remove it later if is a dead end )

I understand xkcd, but that not mean not improve to fix a real problem. 
Just ask new contributors, or fsharp xplat repo ( runnjng a partial test suite because bat/runpl )
Maybe that is not a solution, but we should try to improve. 
My mediumterm goal? Have all suites under nunit. Different implementation ok, but can we at least agree in test runner?

KevinRansom commented:
@enricosada 

Hi,
what is the status of this PR can you make a recommendation about the way forward?  It looks like the feature is blocked on testing.

Thanks

Kevin

KevinRansom commented:
Ping @enricosada  How do you feel about this PR, what is your recommendation?

enricosada commented:
i'll remove this test, and use the nunit test suite, is easier to do that now (call directly the compiler in memory was easier, but it's ok)

enricosada commented:
done, ready to review
added unit tests and some integration tests inside `tests/fsharp` suite

i added `FscCommand.fscToLibrary` helper to easy compile some source files and check warnings, errors and output ( see test `should set file version info on generated file` )

The new helpers use the same functions and config of the `tests/fsharp` suite (so same config for fsc path etc) and run each test in a new temp directory ( so these can be run in parallel without problems ).
 
When we add in memory compilation, we can skip the file system and the test doesnt change much.

All tests can be run and debugged inside visual studio. 

The test output contains:
- the temp test directory
- all command executed ( `fsc test.fs` etc ), so in
- the output if fsc command fails

Commands logged can be rerun from command line inside the temp directory, so really easy to check failures
